src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs

```diff
@@ -55,13 +55,32 @@ private IDictionary CreateDataContainer()
         [MethodImpl(MethodImplOptions.InternalCall)]
         private static extern bool IsImmutableAgileException(Exception e);
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern IRuntimeMethodInfo GetMethodFromStackTrace(object stackTrace);
+        [LibraryImport(RuntimeHelpers.QCall, EntryPoint = "ExceptionNative_GetMethodFromStackTrace")]
+        private static partial void GetMethodFromStackTrace(ObjectHandleOnStack stackTrace, ObjectHandleOnStack method);
 
         private MethodBase? GetExceptionMethodFromStackTrace()
         {
             Debug.Assert(_stackTrace != null, "_stackTrace shouldn't be null when this method is called");
-            IRuntimeMethodInfo method = GetMethodFromStackTrace(_stackTrace!);
+
+            // The _stackTrace can be either sbyte[] or Object[]. In the latter case,
```

Comment:
The object[] array is there to keep the methods referenced in the stack trace from unloading. We need to keep the `object[]` for as long as we are operating on the `byte[]` array. I think we need `GC.KeepAlive` after the QCall.

(The existing FCall implementation has the same bug.)

Reply:
Also, the stacktrace is updated via a lock-free algorithm. The unmanaged `ExceptionObject::GetStackTrace` method is the workhorse of this lock-free algorithm. To avoid race conditions, it creates a local copy of the stacktrace if there may be other threads modifying it. I suspect that the unwrapping of the stacktrace here may have subtle race conditions because of it is not using `ExceptionObject::GetStackTrace`. If you believe that it is not these, we need a comment that explains it.

It would be easier to just use the `ExceptionObject::GetStackTrace` for this one as well (ie unwrap the stacktrace inside the QCall using `ExceptionObject::GetStackTrace` instead of here). No need to worry about the race conditions.

Reply:
cc @janvorli

Reply:
Ah, I wasn't sure why it was `object[]` or `sbyte[]`. The reasoning here makes sense. I will add that.

Reply:
I believe we need to use the `ExceptionObject::GetStackTrace` to be sure that the method on the first frame doesn't get collected

Reply:
Also, when you make that change, the case when stackTrace.Size() is zero would be a valid case.

Reply:
> The unmanaged ExceptionObject::GetStackTrace method is the workhorse of this lock-free algorithm. To avoid race conditions, it creates a local copy of the stacktrace if there may be other threads modifying it.

The `_stacktrace` was set via the result of `GetFrozenStackTrace()`, see [`CaptureDispatchState()`](https://github.com/dotnet/runtime/blob/7cdd2e90db3bf9244c895ecb1518d1eab87e3598/src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs#L239-L245). Do we really need to do this again?

Reply:
> Also, when you make that change, the case when stackTrace.Size() is zero would be a valid case.

I don't see how this can be valid when calling `ExceptionObject::GetStackTrace()` directly, but not when we are using the same result from `GetFrozenStackTrace()`. I think I'm missing something fundamental here.

Reply:
actually, I was wrong, the stackTrace.Size() being zero cannot happen. I was thinking it could happen due to the following if the first frame was a method that can be collected:
https://github.com/dotnet/runtime/blob/7cdd2e90db3bf9244c895ecb1518d1eab87e3598/src/coreclr/vm/object.cpp#L1969-L1976
However, when the first frame is added, this cannot happen, as the stackTrace was null before and adding the frame writes the object[] there that already has the keepalive element for the first stack frame.

Reply:
> The `_stacktrace` was set via the result of `GetFrozenStackTrace()`, see [`CaptureDispatchState()`](https://github.com/dotnet/runtime/blob/7cdd2e90db3bf9244c895ecb1518d1eab87e3598/src/coreclr/System.Private.CoreLib/src/System/Exception.CoreCLR.cs#L239-L245). Do we really need to do this again?

In that case you don't, there is no race possible in this case.

Reply:
Thanks @janvorli. I will confirm this is always the case.

Reply:
From the looks of things the `GetFrozenStackTrace()` is the canonical path.